src/lib/server/endpoints/images.ts

```diff
@@ -0,0 +1,214 @@
+	defaults: ImageProcessorOptions<MimeType>
+) {
+	return z
+		.object({
+			supportedMimeTypes: z
+				.array(
+					z.enum<string, [MimeType, ...MimeType[]]>([
+						defaults.supportedMimeTypes[0],
+						...defaults.supportedMimeTypes.slice(1),
+					])
+				)
+				.default(defaults.supportedMimeTypes),
+			preferredMimeType: z
+				.enum<string, [MimeType, ...MimeType[]]>([
+					defaults.supportedMimeTypes[0],
+					...defaults.supportedMimeTypes.slice(1),
+				])
+				.default(defaults.preferredMimeType),
+			// The 4 / 3 compensates for the 33% increase in size when converting to base64
+			maxSizeInMB: z.number().positive().default(defaults.maxSizeInMB),
+			maxWidth: z.number().int().positive().default(defaults.maxWidth),
+			maxHeight: z.number().int().positive().default(defaults.maxHeight),
+		})
+		.default(defaults);
+}
+
+export interface ImageProcessorOptions<MimeType extends string = string> {
+	supportedMimeTypes: MimeType[];
+	preferredMimeType: MimeType;
+	maxSizeInMB: number;
+	maxWidth: number;
+	maxHeight: number;
+}
+export type ImageProcessor<MimeType extends string = string> = (file: MessageFile) => Promise<{
+	image: Buffer;
+	mime: MimeType;
+}>;
+export function makeImageProcessor<MimeType extends string = string>(
+	options: ImageProcessorOptions<MimeType>
+): ImageProcessor<MimeType> {
+	return async (file) => {
+		const { supportedMimeTypes, preferredMimeType, maxSizeInMB, maxWidth, maxHeight } = options;
+		const { mime, value } = file;
+
+		const buffer = Buffer.from(value, "base64");
+		let sharpInst = sharp(buffer);
+
+		const metadata = await sharpInst.metadata();
+		if (!metadata) throw Error("Failed to read image metadata");
+		const { width, height } = metadata;
+		if (width === undefined || height === undefined) throw Error("Failed to read image size");
+
+		const tooLargeInSize = width > maxWidth || height > maxHeight;
+		const tooLargeInBytes = buffer.byteLength > maxSizeInMB * 1024 * 1024;
+
+		const outputMime = chooseMimeType(supportedMimeTypes, preferredMimeType, mime, {
+			preferSizeReduction: tooLargeInBytes,
+		});
+
+		// Resize if necessary
+		if (tooLargeInSize || tooLargeInBytes) {
+			const size = chooseImageSize({
+				mime: outputMime,
+				width,
+				height,
+				maxWidth,
+				maxHeight,
+				maxSizeInMB,
+			});
+			if (size.width !== width || size.height !== height) {
+				sharpInst = resizeImage(sharpInst, size.width, size.height);
+			}
+		}
+
+		// Convert format if necessary
+		// We always want to convert the image when the file was too large in bytes
+		// so we can guarantee that ideal options are used, which are expected when
+		// choosing the image size
+		if (outputMime !== mime || tooLargeInBytes) {
+			sharpInst = convertImage(sharpInst, outputMime);
+		}
+
+		const processedImage = await sharpInst.toBuffer();
+		return { image: processedImage, mime: outputMime };
+	};
+}
+
+type OutputFormat = "png" | "jpeg" | "webp" | "avif" | "tiff" | "gif";
```

Comment:
```suggestion
type OutputImgFormat = "png" | "jpeg" | "webp" | "avif" | "tiff" | "gif";
```

more explicit file name